### ISSUE TITLE
Handle windows-style bare specifiers in exports

### ISSUE THREAD

**Issue by jkrems** [nodejs/modules#356]
There's currently a loophole in CJS exports:

```js
require('x\\y.js');
```

The above will load `y.js` at least on Windows even if `x` uses `exports` and didn't whitelist the path. This happens because the specifier isn't recognized as an `exports` specifier.

There's three possible ways out:

1. We detect this case and iff `x` uses `exports` treat it as "not mapped", no matter what the subpath is.
2. We accept this as a loophole.
3. We normalize the specifier, turning backslashes into slashes.

There are other potential loopholes like the following:

* `@scope/../any/file.js`: Doesn't match the RegExp, allows access to `any/file.js`.
* TBD

It's not quite clear if we *can* plug all holes in CJS but there's likely some way of doing it.

**Comment by zackschuster** [comment]
i would think normalizing the specifier to the platform would be the expected behavior?

**Comment by jkrems** [comment]
Yes-ish. The problem is that we're crossing streams. It's the same configuration for both ESM and CJS. ESM operates on file URLs so "normalizing" means converting those to platform file system paths. CJS operates on file system paths so "normalizing" means first mapping relative path segments to URL segments, then mapping, and then mapping the result back to file system paths. The question is less "should we normalize" and more "what do we consider our input/output formats to be" so we can normalize correctly.

**Comment by bmeck** [comment]
2. is the general approach wrt some things like case insensitive filesystems using `require()`. Seems fine.

As long as it is documented, I don't have a preference on solution chosen.

If we do go with 3. however I would suggest the following types and coercions:

1. `exports` key is a relative URL string with leading `.` or `..`, or absolute URL string
2. `exports` value is a relative URL string string with leading `.` or `..`, or absolute URL string

Conversion happens lazily like most things. People can use auditing like linters to do ahead of time checks.

The URLs are resolved against the location of the `package.json` file using the regular URL parser:

```mjs
// will ensure trailing / and remove fragment and query
// don't need this if you aren't checking location after resolve
pkgRoot = new URL('.', fileURLOfPackageJSON);
resolvedExportValue = new URL(value, pkgRoot);

// could do a pathname check to see if it stayed in the pkgRoot?
// idk, seems more of an auditing problem than a need to prevent this behavior

// only used by CJS
// this strips query/fragment but CJS doesn't support those to begin with
filename = fileURLToPath(resolvedExportValue);
```

[`fileURLToPath`](https://github.com/nodejs/node/blob/908292cf1f551c614a733d858528ffb13fb3a524/lib/internal/url.js#L1280) handles some minor oddities when converting between things and `new URL` handles `\\` normalization in a standardized way as seen with `new URL('.\\x\\y', 'file:////app/')`.

**Comment by jkrems** [comment]
I think my current thinking is that we should officially narrow the scope of `exports` to what it is right now: specifiers that are a package name or a package name followed by a slash.

This will mean that code can "exploit" backslash on Windows (under `require`) and code can "exploit" case-differences on case-insensitive file systems. But portable code will be using `exports`. Since we can close the more general escape hatch (case-insensitive file systems), it feels weird to add complexity to close a very platform specific one. I don't think I've seen backslash widely used in imports or requires.

**Comment by jkrems** [comment]
Yep. Or to be more precise: when it's a "package name specifier" which is a subset of all bare specifiers. E.g. `\0%$!` is a bare specifier but not something that would trigger exports (a "package name").

**Comment by MylesBorins** [comment]
Since this type of specifier resolution would affect both CJS and ESM would
it make sense for us to transfer this issue to nodejs/node?
